// next-auth/utils/node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBody = getBody;
exports.getURL = getURL;
exports.setCookie = setCookie;
exports.setHeaders = setHeaders;

function setCookie(res, value) {
  var _res$getHeader;

  let setCookieHeader = (_res$getHeader = res.getHeader("Set-Cookie")) !== null && _res$getHeader !== void 0 ? _res$getHeader : [];

  if (!Array.isArray(setCookieHeader)) {
    setCookieHeader = [setCookieHeader];
  }

  setCookieHeader.push(value);
  res.setHeader("Set-Cookie", setCookieHeader);
}

function getBody(req) {
  if (!("body" in req) || !req.body || req.method !== "POST") {
    return;
  }

  if (req.body instanceof ReadableStream) {
    return {
      body: req.body
    };
  }

  return {
    body: JSON.stringify(req.body)
  };
}

function getURL(url, trusted = !!(() => {
  var _process$env$AUTH_TRU;

  return (_process$env$AUTH_TRU = process.env.AUTH_TRUST_HOST) !== null && _process$env$AUTH_TRU !== void 0 ? _process$env$AUTH_TRU : process.env.VERCEL;
})(), forwardedValue) {
  try {
    var _process$env$NEXTAUTH;

    let host = (_process$env$NEXTAUTH = process.env.NEXTAUTH_URL) !== null && _process$env$NEXTAUTH !== void 0 ? _process$env$NEXTAUTH : process.env.NODE_ENV !== "production" && "http://localhost:3000";

    if (trusted && forwardedValue) {
      host = Array.isArray(forwardedValue) ? forwardedValue[0] : forwardedValue;
    }

    if (!host) throw new TypeError("Invalid host");
    if (!url) throw new TypeError("Invalid URL, cannot determine action");

    if (host.startsWith("http://") || host.startsWith("https://")) {
      return new URL(`${host}${url}`);
    }

    return new URL(`https://${host}${url}`);
  } catch (error) {
    return error;
  }
}

function getSetCookies(cookiesString) {
  if (typeof cookiesString !== "string") {
    return [];
  }

  const cookiesStrings = [];
  let pos = 0;
  let start;
  let ch;
  let lastComma;
  let nextStart;
  let cookiesSeparatorFound;

  function skipWhitespace() {
    while (pos < cookiesString.length && /\s/.test(cookiesString.charAt(pos))) {
      pos += 1;
    }

    return pos < cookiesString.length;
  }

  function notSpecialChar() {
    ch = cookiesString.charAt(pos);
    return ch !== "=" && ch !== ";" && ch !== ",";
  }

  while (pos < cookiesString.length) {
    start = pos;
    cookiesSeparatorFound = false;

    while (skipWhitespace()) {
      ch = cookiesString.charAt(pos);

      if (ch === ",") {
        lastComma = pos;
        pos += 1;
        skipWhitespace();
        nextStart = pos;

        while (pos < cookiesString.length && notSpecialChar()) {
          pos += 1;
        }

        if (pos < cookiesString.length && cookiesString.charAt(pos) === "=") {
          cookiesSeparatorFound = true;
          pos = nextStart;
          cookiesStrings.push(cookiesString.substring(start, lastComma));
          start = pos;
        } else {
          pos = lastComma + 1;
        }
      } else {
        pos += 1;
      }
    }

    if (!cookiesSeparatorFound || pos >= cookiesString.length) {
      cookiesStrings.push(cookiesString.substring(start, cookiesString.length));
    }
  }

  return cookiesStrings;
}

function setHeaders(headers, res) {
  for (const [key, val] of headers.entries()) {
    const value = key === "set-cookie" ? getSetCookies(val) : val;
    res.setHeader(key, value);
  }
}
