// next-auth/utils/web.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toInternalRequest = toInternalRequest;
exports.toResponse = toResponse;

var _cookie = require("cookie");

const decoder = new TextDecoder();

async function streamToString(stream) {
  const chunks = [];
  return await new Promise((resolve, reject) => {
    stream.on("data", chunk => chunks.push(Buffer.from(chunk)));
    stream.on("error", err => reject(err));
    stream.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
  });
}

async function readJSONBody(body) {
  try {
    if ("getReader" in body) {
      const reader = body.getReader();
      const bytes = [];

      while (true) {
        const {
          value,
          done
        } = await reader.read();
        if (done) break;
        bytes.push(...value);
      }

      const b = new Uint8Array(bytes);
      return JSON.parse(decoder.decode(b));
    }

    if (typeof Buffer !== "undefined" && Buffer.isBuffer(body)) {
      return JSON.parse(body.toString("utf8"));
    }

    return JSON.parse(await streamToString(body));
  } catch (e) {
    console.error(e);
  }
}

async function toInternalRequest(req) {
  var _req$headers$get, _parseCookie, _url$searchParams$get;

  const url = new URL(req.url);
  const nextauth = url.pathname.split("/").slice(3);
  const headers = Object.fromEntries(req.headers);
  const query = Object.fromEntries(url.searchParams);
  const cookieHeader = (_req$headers$get = req.headers.get("cookie")) !== null && _req$headers$get !== void 0 ? _req$headers$get : "";
  const cookies = (_parseCookie = (0, _cookie.parse)(Array.isArray(cookieHeader) ? cookieHeader.join(";") : cookieHeader)) !== null && _parseCookie !== void 0 ? _parseCookie : {};
  return {
    action: nextauth[0],
    method: req.method,
    headers,
    body: req.body ? await readJSONBody(req.body) : undefined,
    cookies: cookies,
    providerId: nextauth[1],
    error: (_url$searchParams$get = url.searchParams.get("error")) !== null && _url$searchParams$get !== void 0 ? _url$searchParams$get : undefined,
    host: new URL(req.url).origin,
    query
  };
}

function toResponse(res) {
  var _res$headers, _res$cookies, _res$status;

  const headers = new Headers((_res$headers = res.headers) === null || _res$headers === void 0 ? void 0 : _res$headers.reduce((acc, {
    key,
    value
  }) => {
    acc[key] = value;
    return acc;
  }, {}));
  (_res$cookies = res.cookies) === null || _res$cookies === void 0 ? void 0 : _res$cookies.forEach(cookie => {
    const {
      name,
      value,
      options
    } = cookie;
    const cookieHeader = (0, _cookie.serialize)(name, value, options);

    if (headers.has("Set-Cookie")) {
      headers.append("Set-Cookie", cookieHeader);
    } else {
      headers.set("Set-Cookie", cookieHeader);
    }
  });
  const body = headers.get("content-type") === "application/json" ? JSON.stringify(res.body) : res.body;
  const response = new Response(body, {
    headers,
    status: res.redirect ? 302 : (_res$status = res.status) !== null && _res$status !== void 0 ? _res$status : 200
  });

  if (res.redirect) {
    response.headers.set("Location", res.redirect);
  }

  return response;
}
